packages/runtime/container-runtime/src/batchTracker.ts

```diff
@@ -0,0 +1,82 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ChildLogger } from "@fluidframework/telemetry-utils";
+import EventEmitter from "events";
+
+export class BatchTracker {
+    private readonly logger: ITelemetryLogger;
+    private startBatchSequenceNumber: number | undefined;
+    private trackedBatchCount: number = 0;
+    private batchProcessingStartTimeStamp: number | undefined;
+
+    constructor(
+        private readonly batchEventEmitter: EventEmitter,
+        logger: ITelemetryLogger,
+        private readonly opCountThreshold: number,
+        private readonly batchCountSamplingRate: number,
+        private readonly dateTimeProvider: () => number = () => Date.now(),
+    ) {
+        this.logger = ChildLogger.create(logger, "Batching");
+
+        this.batchEventEmitter.on("batchBegin", (message: ISequencedDocumentMessage) => {
+            this.startBatchSequenceNumber = message.sequenceNumber;
+            this.batchProcessingStartTimeStamp = this.dateTimeProvider();
+            this.trackedBatchCount++;
+        });
+
+        this.batchEventEmitter.on("batchEnd", (error: any | undefined, message: ISequencedDocumentMessage) => {
+            assert(
+                this.startBatchSequenceNumber !== undefined && this.batchProcessingStartTimeStamp !== undefined,
+                "batchBegin must fire before batchEnd");
+
+            const opCount = message.sequenceNumber - this.startBatchSequenceNumber + 1;
+            if (opCount >= this.opCountThreshold) {
+                this.logger.sendErrorEvent({
+                    eventName: "TooManyOps",
+                    opCount,
+                    threshold: opCountThreshold,
+                    referenceSequenceNumber: message.referenceSequenceNumber,
+                    batchEndSequenceNumber: message.sequenceNumber,
+                    timeSpanMs: this.dateTimeProvider() - this.batchProcessingStartTimeStamp,
```

Comment:
we call it duration, and logger has some special processing for it (it rounds it to integer, in case it is floating number).
BTW, if you were to use PerformanceEvent, it would do it automatically - only _end event would be recorded (with duration) by default.

Reply:
same below